### PR TITLE
Use flags for cache settings

### PR DESCRIPTION
**Pull request by tempoz**
Use Register functions to move implementation details out of main functions.

---

**Version bump**: Minor

### CODE REVIEW

**Comment by tempoz** [comment]
Current dependencies on/for this PR:
* master
  * **PR #1771** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/1771" target="_blank">Graphite</a> 
    * **PR #1772** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/1772" target="_blank">Graphite</a> 
      * **PR #1773** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/1773" target="_blank">Graphite</a> 
        * **PR #1774** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/1774" target="_blank">Graphite</a> 
          * **PR #1775** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/1775" target="_blank">Graphite</a> 
            * **PR #1776** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/1776" target="_blank">Graphite</a> 
              * **PR #1777** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/1777" target="_blank">Graphite</a> 
                * **PR #1778** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/1778" target="_blank">Graphite</a> 
                  * **PR #1779** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/1779" target="_blank">Graphite</a> 
                    * **PR #1780** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/1780" target="_blank">Graphite</a> 
                      * **PR #1783** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/1783" target="_blank">Graphite</a> 
                      * **PR #1833** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/1833" target="_blank">Graphite</a> 
                        * **PR #1834** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/1834" target="_blank">Graphite</a> 
                          * **PR #1835** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/1835" target="_blank">Graphite</a> 
                            * **PR #1836** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/1836" target="_blank">Graphite</a>  (this PR)
                              * **PR #1837** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/1837" target="_blank">Graphite</a> 
                                * **PR #1838** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/1838" target="_blank">Graphite</a> 
                                  * **PR #1839** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/1839" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/1836?utm_source=stack-comment).

File: enterprise/server/backends/memcache/memcache.go
Comment (tylerwilliams) [inline enterprise/server/backends/memcache/memcache.go:46]: should this log a warning? does this indicate a misconfiguration or logic error in our code?

File: enterprise/server/backends/memcache/memcache.go
Comment (tylerwilliams) [inline enterprise/server/backends/memcache/memcache.go:51]: nit: move up first

File: enterprise/server/backends/memcache/memcache.go
Comment (tylerwilliams) [inline enterprise/server/backends/memcache/memcache.go:45]: nit: maybe we should export composableCache.ComposableCache so that we can do these checks without resorting to stringification.

Then we can just write:

if _, ok := env.GetCache().(*composable_cache.ComposableCache); ok {
  // warning time
}

File: enterprise/server/raft/cache/cache.go
Comment (tylerwilliams) [inline enterprise/server/raft/cache/cache.go:115]: defers are executed right before the function returns, so this will actually stop the cache before Register returns. I don't think that's what we want. The reason this was OK in main before is because main loops forever until exit.

You could instead register a graceful shutdown function (using the healthchecker in the env) that calls rc.Stop?

**Review by tylerwilliams — CHANGES_REQUESTED** [review]

File: enterprise/server/raft/cache/cache.go
Comment (tempoz) [reply enterprise/server/raft/cache/cache.go:116]: Thanks for the catch, fixed!

File: enterprise/server/backends/memcache/memcache.go
Comment (tempoz) [reply enterprise/server/backends/memcache/memcache.go]: Done.

File: enterprise/server/backends/memcache/memcache.go
Comment (tempoz) [reply enterprise/server/backends/memcache/memcache.go]: Previously, we allowed both memcache and redis cache to be configured and memcache would silently take precedence if present, so these checks are intended to just duplicate that behavior by only allowing one composable cache to be used and, using the first one we try to register that is also configured. If we think this is confusing and want to log a warning, I have no objections, but it is a departure from the current way the system works.

File: enterprise/server/backends/memcache/memcache.go
Comment (tempoz) [reply enterprise/server/backends/memcache/memcache.go]: Done! and did the analogous thing in redis_cache.

File: enterprise/server/backends/memcache/memcache.go
Comment (tylerwilliams) [inline enterprise/server/backends/memcache/memcache.go:48]: nit: log a warning in this case because it indicates a misconfiguration?

File: enterprise/server/backends/redis_cache/redis_cache.go
Comment (tylerwilliams) [inline enterprise/server/backends/redis_cache/redis_cache.go:52]: nit: same as above, can we also log a warning here so that people can fix their flags?

File: enterprise/server/cmd/executor/executor.go
Comment (tylerwilliams) [inline enterprise/server/cmd/executor/executor.go:144]: log.Fatal(err) (here and below)

File: enterprise/server/cmd/server/main.go
Comment (tylerwilliams) [inline enterprise/server/cmd/server/main.go:184]: nit: same as above, here and below

File: enterprise/server/raft/cache/cache.go
Comment (tylerwilliams) [inline enterprise/server/raft/cache/cache.go:116]: Looks like defer rc.Stop() is still here (L116?)

File: server/backends/disk_cache/disk_cache.go
Comment (tylerwilliams) [inline server/backends/disk_cache/disk_cache.go:170]: nit: i think this formatting is kinda weird -- maybe cleaner as:

dc := &config.DiskConfig{
			RootDirectory:     *rootDirectory,
			Partitions:        partitions,
			PartitionMappings: partitionMappings,
			UseV2Layout:       *useV2Layout,
		}
c, err := NewDiskCache(env, dc, cache_config.MaxSizeBytes())

File: server/backends/memory_cache/memory_cache.go
Comment (tylerwilliams) [inline server/backends/memory_cache/memory_cache.go:42]: I think we should always warn people when we detect what we'd consider a "misconfiguration" like two caches being configured.

It's sort of preventative medicine in the sense that it will help our customers figure out their own misconfigurations and fix them before they have to reach out to us and ask "how to i setup bb".

File: server/libmain/libmain.go
Comment (tylerwilliams) [inline server/libmain/libmain.go:167]: same as above. (anytime you see a single verb formatting string like this that's not doing a type conversion it's almost always a sign that the code could be simplified)

**Review by tylerwilliams — CHANGES_REQUESTED** [review]

**Review by tylerwilliams — APPROVED** [review]

**Comment by tempoz** [comment]
**Graphite Merge Job**

Current status: ⏳ Queued to merge

This pull request is currently queued to be merged as part of a stack.

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/1836?utm_source=stack-comment).


Job Reference: LYGXjCZ9jzXE0ra5XapW

**Comment by tempoz** [comment]
**Graphite Merge Job**

Current status: ⏳ Queued to merge

This pull request is currently queued to be merged as part of a stack.

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/1836?utm_source=stack-comment).


Job Reference: bO0bzB7R9g7bK8rQ9uPT

**Comment by tempoz** [comment]
**Graphite Merge Job**

Current status: ⏳ Queued to merge

This pull request is currently queued to be merged as part of a stack.

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/1836?utm_source=stack-comment).


Job Reference: x9vSQMZuwVwdzOiJU1Ph

**Comment by tempoz** [comment]
**Graphite Merge Job**

Current status: ⏳ Queued to merge

This pull request is currently queued to be merged as part of a stack.

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/1836?utm_source=stack-comment).


Job Reference: xLhUv52COpx6k7AsRAsn

**Comment by tempoz** [comment]
**Graphite Merge Job**

Current status: ⏳ Queued to merge

This pull request is currently queued to be merged as part of a stack.

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/1836?utm_source=stack-comment).


Job Reference: KZCXEHPsl7FVcdrah9Au

**Comment by tempoz** [comment]
**Graphite Merge Job**

Current status: 🔄 Waiting for checks to pass

This pull request is currently waiting for CI and mergeability checks to pass before being merged.

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/1836?utm_source=stack-comment).


Job Reference: eJKjgKqVU3dUX0WhID3h

**Pull request merged**
